### ISSUE TITLE
Flexbits mmap staying active after expired

### ISSUE THREAD

**Issue by bryant-smith** [quadrantsec/sagan#54]
There is an issue with the countdown for mmap.  When I use flexbits/xbits and monitor the variables using Saganpeek, I've watched them go to zero and then negative but without it going INACTIVE.  They stay active and affect the rules ability to trigger correctly.

```
Flexbit name: "SMB.IPC.Request"
State: ACTIVE
IP: 192.168.44.139:514 -> [REDACTED]:514
Username: ""
Signature: "[WINDOWS-SECURITY] SMB IPC Share Request" (Signature ID: 1111111)
Expire Time: Jan 25 05:22:16 2023 (10 seconds)
Time until expire: -8 seconds.
Syslog message: " 5145: A network share object was checked to see whether client can be granted desired access. Subject: Security ID: S-1-5-21-3178379434-3483718794-4062958107-1106 Account Name: GRYFFINDOR$ Account Domain: HOGWARTS Logon ID: 0x183B53 Network Information: Object Type: File Source Address: 192.168.44.139 Source Port: 58263 Share Information: Share Name: \\*\IPC$ Share Path: Relative Target Name: samr Access Request Information: Access Mask: 0x12019F Accesses: READ_CONTROL SYNCHRONIZE ReadData (or ListDirectory) WriteData (or AddFile) AppendData (or AddSubdirectory or CreatePipeInstance) ReadEA WriteEA ReadAttributes WriteAttributes Access Check Results: - "

Type: flexbit [2].
Flexbit name: "SMB.C.Request"
State: ACTIVE
IP: 192.168.44.139:514 -> [REDACTED]:514
Username: ""
Signature: "[WINDOWS-SECURITY] SMB C Share Request" (Signature ID: 1111113)
Expire Time: Jan 25 05:22:16 2023 (10 seconds)
Time until expire: -8 seconds.
Syslog message: " 5145: A network share object was checked to see whether client can be granted desired access. Subject: Security ID: S-1-5-21-3178379434-3483718794-4062958107-500 Account Name: Administrator Account Domain: HOGWARTS Logon ID: 0x17AB33 Network Information: Object Type: File Source Address: 192.168.44.139 Source Port: 58526 Share Information: Share Name: \\*\C$ Share Path: \??\C:\ Relative Target Name: \ Access Request Information: Access Mask: 0x100081 Accesses: SYNCHRONIZE ReadData (or ListDirectory) ReadAttributes Access Check Results: - "

Type: flexbit [3].
Flexbit name: "SMB.ADMIN.Request"
State: ACTIVE
IP: 192.168.44.139:514 -> [REDACTED]514
Username: ""
Signature: "[WINDOWS-SECURITY] SMB ADMIN Share Request" (Signature ID: 1111112)
Expire Time: Jan 25 05:22:16 2023 (10 seconds)
Time until expire: -8 seconds.
Syslog message: " 5145: A network share object was checked to see whether client can be granted desired access. Subject: Security ID: S-1-5-21-3178379434-3483718794-4062958107-500 Account Name: Administrator Account Domain: HOGWARTS Logon ID: 0x17AB33 Network Information: Object Type: File Source Address: 192.168.44.139 Source Port: 58526 Share Information: Share Name: \\*\ADMIN$ Share Path: \??\C:\Windows Relative Target Name: \ Access Request Information: Access Mask: 0x100081 Accesses: SYNCHRONIZE ReadData (or ListDirectory) ReadAttributes Access Check Results: - "
```

**Comment by CyberTaoFlow** [comment]
I have noticed this as well

**Comment by quadrantsec** [comment]
I looked into this and believe it or not,  flexbits are working as normal.  I think I need to redesign "saganpeek" to better represent what is happening!

Basically,  there is an internal flag (flexbit_ipc[i].flexbit_state) that Sagan keeps track of.   "saganpeek" is a "read only" tool and does not alter or change the state. 

Imagine you set a flexbit for 30 seconds and let timer go paste 30 seconds.   When you run "saganpeek",  it displays all "ACTIVE" flexbits and the timer will be negative.   It's "ACTIVE" because "sagan" hasn't updated the flexbit_ipc[i].flexbit_state.  The flexbit_ipc[i].flexbit_state only changes when "sagan "touches that flexbit.  

So, if you're flexbit is past 30 seconds and you do a check via "flexbit isset",  "sagan" is now accessing that memory.  As part of its routine,  it will set the flexbit_ipc[i].flexbit_state to "false".   However,  until (or if!) that call happens,  the flexbit_ipc[i].flexbit_state will remain true.  Hence the negative timer and "ACTIVE" flexbit!

Long story short,  flexbits are working as normal but "saganpeek" is showing it in a really dumb way.  I'm thinking I might alter the tool to say: 

ACTIVE 
INACTIVE

and if the timer is negative,  

TIMEOUT 

and _not_ display it.  This means,  in order to see "timeout" or "inactive" flexbits, you'd have to pass the --all to "saganpeek". 

hopefully this makes some sense. 

I'll close this when I've modified saganpeek.

**Comment by bryant-smith** [comment]
This makes perfect sense and good to know.  TIMEOUT may be the better option if the value is negative but still "active" until Sagan checks the flexbit_state.  I guess this only makes sense if and when the memory location is checked and set to false you can run saganpeek --all and it will still show up.

**Comment by quadrantsec** [comment]
I've patch this.  If the timer is negative, it no longer displays.   If the -a is used,  it will show "TIMEOUT" instead of active.  Hopefully less confusion this way!

# ./saganpeek -a
Type: flexbit [0].
Flexbit name: "testme"
State: TIMEOUT
IP: 192.168.2.1:514 -> 192.168.2.1:514
Signature: "Flexbit set" (Signature ID: 5002014)
Expire Time: Feb 17 11:54:07 2023 (30 seconds)
Time until expire: -327 seconds.
Syslog message: " flexbit set"
